### PR TITLE
Passkey: check account signature on passkey

### PR DESCRIPTION
**Pull request by saraswatpuneet**
Go after #2037 

# Goal
The goal of this PR is to setup functions and tests for account signature checks

Part of: #2028 

# Checklist
- [ ] Chain spec updated
- [ ] Custom RPC OR Runtime API added/changed? Updated js/api-augment.
- [ ] Design doc(s) updated
- [x] Tests added
- [ ] Benchmarks added
- [ ] Weights updated

### CODE REVIEW

File: pallets/passkey/src/lib.rs [hunk 1]
```diff
@@ -89,6 +95,34 @@ pub mod module {
...
+			let signed_payload: Vec<u8> = wrap_binary_data(signed_data.clone().into());
+
+			let verified = signature.verify(&signed_payload[..], &key);
+			if verified {
+				Ok(())
+			} else {
+				Err(Error::<T>::InvalidAccountSignature.into())
+			}
+		}
+	}
```
Comment (saraswatpuneet) [inline pallets/passkey/src/lib.rs:125]: this is a private function so not tests are added yet

File: pallets/passkey/src/mock.rs [hunk 2]
```diff
@@ -1,23 +1,24 @@
...
 
-pub type AccountId = u128;
 impl frame_system::Config for Test {
 	type RuntimeOrigin = RuntimeOrigin;
 	type RuntimeCall = RuntimeCall;
 	type Nonce = u64;
 	type Hash = H256;
 	type Hashing = ::sp_runtime::traits::BlakeTwo256;
-	type AccountId = AccountId;
+	type AccountId = u64;
```
Comment (JoeCap08055) [inline pallets/passkey/src/mock.rs:21]: Should we use `crate::msa::MessageSourceId` here instead? I'd prefer not to propagate `u64` in different places if it's all referencing the same conceptual type.

**Review by JoeCap08055 — APPROVED** [review]

File: pallets/passkey/src/mock.rs [hunk 3]
```diff
@@ -1,23 +1,24 @@
...
 
-pub type AccountId = u128;
 impl frame_system::Config for Test {
 	type RuntimeOrigin = RuntimeOrigin;
 	type RuntimeCall = RuntimeCall;
 	type Nonce = u64;
 	type Hash = H256;
 	type Hashing = ::sp_runtime::traits::BlakeTwo256;
-	type AccountId = AccountId;
+	type AccountId = MessageSourceId;
```
Comment (aramikm) [inline pallets/passkey/src/mock.rs:21]: AccountId is different from MsaId. These also should be a type that allows us to check signature in tests. In my PR I believe the types which are set allows this.

Comment (saraswatpuneet) [reply pallets/passkey/src/mock.rs:21]: True, got confused , add the correct AccountId

Comment (saraswatpuneet) [reply pallets/passkey/src/mock.rs]: I update mock on my side, Merging after your PR would be ideal

File: pallets/passkey/src/lib.rs [hunk 4]
```diff
@@ -126,6 +141,35 @@ pub mod module {
...
+		) -> DispatchResult {
+			let key: AccountId32 = T::ConvertIntoAccountId32::convert((*signer).clone());
+			let signed_payload: Vec<u8> = wrap_binary_data(signed_data.clone().into());
+
+			let verified = signature.verify(&signed_payload[..], &key);
+			if verified {
+				Ok(())
+			} else {
+				Err(Error::<T>::InvalidAccountSignature.into())
+			}
```
Comment (saraswatpuneet) [inline pallets/passkey/src/lib.rs:170]: This function validated the account signature on passkey public

File: pallets/passkey/src/lib.rs [hunk 5]
```diff
@@ -106,6 +113,14 @@ pub mod module {
...
 			));
+
+			// check the signature
+			Self::check_account_signature(
+				&transaction_account_id,
+				&payload.passkey_public_key.into(),
+				&payload.passkey_call.account_ownership_proof,
+			)?;
+
 			let result = payload.passkey_call.call.dispatch(main_origin);
```
Comment (saraswatpuneet) [inline pallets/passkey/src/lib.rs:124]: Validation: check for valid account ownership

File: pallets/passkey/src/tests.rs [hunk 6]
```diff
@@ -25,7 +28,7 @@ fn proxy_call_with_signed_origin_should_fail() {
 			})),
 		};
 		let payload = PasskeyPayload {
-			passkey_public_key: [0u8; 33],
+			passkey_public_key,
 			verifiable_passkey_signature: VerifiablePasskeySignature {
```
Comment (saraswatpuneet) [inline pallets/passkey/src/tests.rs:32]: Modified tests to pass correct data so signature is correct

**Review by mattheworris — APPROVED** [review]
- Read through code
- Executed 'unit tests' (all passed)
🚀 lgtm 🚀

File: pallets/passkey/src/lib.rs [hunk 7]
```diff
@@ -106,6 +113,14 @@ pub mod module {
 			let main_origin = T::RuntimeOrigin::from(frame_system::RawOrigin::Signed(
 				transaction_account_id.clone(),
 			));
+
+			// check the signature
+			Self::check_account_signature(
```
Comment (aramikm) [inline pallets/passkey/src/lib.rs:118]: This should be in called in validate. I think having this check in here is duplicating the same check.

File: pallets/passkey/src/lib.rs [hunk 8]
```diff
@@ -126,6 +141,35 @@ pub mod module {
...
+		/// Returns Ok(()) if the signature is valid
+		/// Returns Err(InvalidAccountSignature) if the signature is invalid
+		/// # Arguments
+		/// * `signer` - The account id of the signer
+		/// * `signed_data` - The signed data
+		/// * `signature` - The signature
+		/// # Return
+		/// * `Ok(())` if the signature is valid
+		/// * `Err(InvalidAccountSignature)` if the signature is invalid
+		#[allow(unused)]
```
Comment (aramikm) [inline pallets/passkey/src/lib.rs:156]: nit: why do we need this?

File: pallets/passkey/src/lib.rs (hunk 7 above)
Comment (saraswatpuneet) [reply pallets/passkey/src/lib.rs:118]: Yes! I forgot about that, will move it in validate

File: pallets/passkey/src/lib.rs (hunk 8 above)
Comment (saraswatpuneet) [reply pallets/passkey/src/lib.rs:156]: Not anymore, will remove it

File: pallets/passkey/src/lib.rs (hunk 7 above)
Comment (saraswatpuneet) [reply pallets/passkey/src/lib.rs]: Done

**Review by aramikm — APPROVED** [review]

**Pull request merged**